### PR TITLE
[DistillationTrainer refactor] Fix `num_items_in_batch` to count generated completion tokens ⚠️ changes loss values

### PR DESCRIPTION
**Pull request by qgallouedec**
*Plan step 3 (re-derived) — stacked on PR 13. First PR of the re-ordered Phase 3. Part of #6449.*

The loss denominator was wrong. `transformers.Trainer` computes `num_items_in_batch` from the **raw dataloader** labels — *before* `_prepare_inputs` replaces the completions with the student's on-policy generations. So the count is wrong for on-policy training (it counts the dataset's completions, not the generated ones), and **zero for prompt-only datasets** (no dataset completions → `num_items = 0` → loss `= sum/0 = NaN`, poisoning the model). This is why the two num_items tests were `xfail`.

This is the fix the re-derived plan moves to the **front** (was plan 5.6): a working `num_items` is the prerequisite for the whole prompt-only dataset contract — you can't switch tests/docs/example to prompt-only (item 15) or remove `messages` (item 16) while prompt-only training NaNs.

### Trainer
- `_fill_buffer` already generates **every** gradient-accumulation-window slice upfront into `self._buffered_inputs`, so the global generated-completion-token count is known before any loss is computed. Compute it there — `sum((labels != -100))` over the buffered slices, `accelerator.gather`-ed across processes — and store it on `self._buffered_num_items` (gathered-global, matching the semantics `_reduce_divergence_loss` already expects).
- `compute_loss` uses `self._buffered_num_items` as the denominator during training, instead of the transformers arg. Eval is unchanged (no buffer).

### Tests
- Un-xfail `test_num_items_in_batch_counts_the_tokens_trained_on` (item 03) — the applied denominator now equals the completion tokens trained on.
- Un-xfail `test_train_runs_with_prompt_only_dataset` (item 12) — prompt-only training no longer NaNs; parameters stay finite.

### This is superseded later
It's a buffer-path fix (~15 lines). When generation moves to the GRPO-style `_prepare_inputs` (item 29), `num_items` is computed natively there (`_generate_and_score_completions`) and this buffer computation is deleted. Kept small and self-contained for exactly that reason — it unblocks Phase 3 without waiting for the generation rework.

### Verification (pytest)

`test_distillation_trainer.py`: **40 passed** (was 38 passed + 2 xfailed). `test_server_distillation_trainer.py` + `test_iw_opd_trainer.py`: **55 passed** (the server subclass inherits `_fill_buffer` but keeps its own `compute_loss`; the frozen iw_opd snapshot is untouched). ruff clean.

### Risk

Medium — it moves loss values (one of the three flagged behaviour-change PRs, alongside items 35 and 38). The per-element objective is unchanged (still pinned by item 01's `generalized_jsd_loss` test); only the reduction denominator changes. Training tests assert finiteness / parameter movement, not exact loss values, so they stay green.

---

> [!NOTE]
> **Medium Risk**
> Loss magnitudes and effective learning rates change for existing distillation runs because only the reduction denominator changed, not the per-token JSD objective; prompt-only training goes from broken (NaN) to working.
> 
> **Overview**
> **Fixes the loss denominator for on-policy `DistillationTrainer`.** Hugging Face `Trainer` was supplying `num_items_in_batch` from **raw collated labels** (dataset completions), which is wrong once the student generates completions and is **zero** for prompt-only datasets (NaN loss). After generation in `_fill_buffer`, the trainer now sums completion tokens (`labels != -100`) across the grad-accum buffer, **gathers** the count across processes into `_buffered_num_items`, and `compute_loss` uses that (with `clamp(min=1)` and per-process DDP adjustment) when calling `_reduce_divergence_loss`.
> 
> **Trainer loss scaling** is adjusted so normalization stays under trainer control: a dummy `compute_loss_func` disables HF’s default grad-accum loss scaling, and `model_accepts_loss_kwargs` is set to `False`.
> 
> **`ServerDistillationTrainer`** divides training loss by `current_gradient_accumulation_steps` because its `compute_loss` override does not use the buffered global token count.
> 
> **Tests:** `test_train_runs_with_prompt_only_dataset` and `test_num_items_in_batch_counts_the_tokens_trained_on` are no longer `xfail`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5d9178afd905b2b4b0c71ced6f06bfe930ea2c6a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/distillation/distillation_trainer.py [hunk 1]
```diff
@@ -718,6 +719,17 @@ def _fill_buffer(self, generation_batch: dict[str, torch.Tensor | Any], buffer_s
...
+        # Loss denominator (`num_items_in_batch`): the global number of completion tokens actually trained on this
+        # optimizer step. transformers derives its own count from the *raw dataloader* labels — before generation
+        # replaces the completions — which is wrong for on-policy training and zero for prompt-only datasets (dividing
+        # the loss by zero). Recompute it here from the generated labels, gathered across processes (issue #4719).
+        local_completion_tokens = sum(
+            int((s["labels"] != -100).sum()) for s in self._buffered_inputs if s is not None
+        )
+        self._buffered_num_items = self.accelerator.gather(
+            torch.tensor(local_completion_tokens, device=self.accelerator.device)
+        ).sum()
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:740]: Good catch, fixed. compute_loss now divides the global denominator by self.accelerator.num_processes to compensate for DDP gradient averaging, matching GRPO

File: trl/experimental/distillation/distillation_trainer.py [hunk 2]
```diff
@@ -1013,6 +1025,11 @@ def _get_teacher_logits(self, inputs: dict[str, torch.Tensor | Any]) -> torch.Te
             ).logits
 
     def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
+        # transformers computes `num_items_in_batch` from the raw dataloader labels, before on-policy generation
+        # replaces the completions; use the count over the generated completions instead (computed in `_fill_buffer`).
+        if self.model.training and self._buffered_num_items is not None:
+            num_items_in_batch = self._buffered_num_items
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py]: Fixed, set compute_loss_func="non-None value to disable scaling" in super().__init__ so training_step doesn't also scale by gradient_accumulation_steps, same mechanism as GRPO

File: trl/experimental/distillation/distillation_trainer.py [hunk 3]
```diff
@@ -486,6 +486,12 @@ def __init__(
             callbacks=callbacks,
             optimizers=optimizers,
             preprocess_logits_for_metrics=preprocess_logits_for_metrics,
+            # In Trainer, `training_step` scales the loss by `gradient_accumulation_steps` only if `compute_loss_func`
+            # is None. Here, loss scaling instead depends on the total number of completion tokens across the global
+            # accumulated batch. To control scaling ourselves, we must disable Trainer's built-in scaling. The simplest
+            # (though a bit hacky) way is to set `compute_loss_func` to any non-None value, which bypasses that behavior
+            # without rewriting `training_step`.
+            compute_loss_func="non-None value to disable scaling",
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:494]: - Added `self.model_accepts_loss_kwargs = False` right after `super().__init__`, verbatim from GRPO/DPO/KTO/RLOO. Without it Trainer's token-averaging path stays active and can compound the /num_processes normalization.
- The `ServerDistillationTraine`r regression is real: it normalizes with batchmean and doesn't consume num_items_in_batch, so disabling grad-accum scaling in the base left it unscaled. Fixed by re-applying loss = loss / self.current_gradient_accumulation_steps in its compute_loss (guarded on self.model.training so eval loss is untouched).

File: trl/experimental/distillation/distillation_trainer.py [hunk 4]
```diff
@@ -718,6 +730,15 @@ def _fill_buffer(self, generation_batch: dict[str, torch.Tensor | Any], buffer_s
         # Generate student completions for every slice
         self._generate_student_completions(slices, list(range(buffer_steps)))
 
+        # Loss denominator (`num_items_in_batch`): the global number of completion tokens actually trained on this
+        # optimizer step. transformers derives its own count from the *raw dataloader* labels — before generation
+        # replaces the completions — which is wrong for on-policy training and zero for prompt-only datasets (dividing
+        # the loss by zero). Recompute it here from the generated labels, gathered across processes (issue #4719).
+        local_completion_tokens = sum(int((s["labels"] != -100).sum()) for s in self._buffered_inputs if s is not None)
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:737]: Not a bug. .sum() returns a 0-dim (scalar) tensor, and int() on a scalar tensor calls .item() implicitly; this works on CUDA (with a sync), it does not raise. The TypeError you're describing only occurs on multi-element tensors, which .sum() never produces. Verified on GPU: `int((labels != -100).sum())` returns the count with no error.

**Review by cmpatino — APPROVED** [review]
LGTM!
